File: src/agenteval/runner.py
"""Agent runner — executes coding agents against tasks."""

import json
import subprocess
import tempfile
import time
from datetime import datetime
from pathlib import Path

from agenteval.models import RunResult, TaskDefinition

# Default agent command templates.
# {prompt} is replaced with the prompt file path, {workdir} with the working directory.
AGENT_CONFIGS: dict[str, str] = {
    "claude-code": "claude -p {prompt} --output-dir {workdir}",
    "aider": "aider --message {prompt}",
}


class AgentRunner:
    """Runs a coding agent on a task and captures the result."""

    def __init__(
        self,
        agent: str = "claude-code",
        agent_cmd: str | None = None,
        timeout: int = 300,
    ):
        if agent_cmd:
            self.cmd_template = agent_cmd
        elif agent in AGENT_CONFIGS:
            self.cmd_template = AGENT_CONFIGS[agent]
        else:
            raise ValueError(
                f"Unknown agent '{agent}'. Known agents: {list(AGENT_CONFIGS)}. "
                "Or pass --agent-cmd with a custom command template."
            )
        self.agent = agent
        self.timeout = timeout

    def run_task(self, task: TaskDefinition, workdir: Path) -> RunResult:
        """Clone repo at base_commit, run agent, capture diff."""
        started_at = datetime.now()
        start_time = time.monotonic()

        try:
            # Clone repo into workdir
            workdir.mkdir(parents=True, exist_ok=True)
            subprocess.run(
                ["git", "clone", task.repo_url, str(workdir / "repo")],
                capture_output=True, text=True, check=True, timeout=120,
            )
            repo_dir = workdir / "repo"
            subprocess.run(
                ["git", "checkout", task.base_commit],
                capture_output=True, text=True, check=True,
                cwd=str(repo_dir), timeout=30,
            )

            # Write prompt file
            prompt_file = workdir / "prompt.txt"
            prompt_file.write_text(task.prompt)

            # Build command
            cmd = self.cmd_template.format(
                prompt=str(prompt_file),
                workdir=str(repo_dir),
            )

            # Run agent
            proc = subprocess.run(
                cmd, shell=True, capture_output=True, text=True,
                cwd=str(repo_dir), timeout=self.timeout,
            )

            # Capture diff
            diff_proc = subprocess.run(
                ["git", "diff"],
                capture_output=True, text=True,
                cwd=str(repo_dir), timeout=30,
            )
            # Also include untracked files
            untracked_proc = subprocess.run(
                ["git", "diff", "--cached"],
                capture_output=True, text=True,
                cwd=str(repo_dir), timeout=30,
            )
            actual_diff = diff_proc.stdout + untracked_proc.stdout

            elapsed = time.monotonic() - start_time
            completed_at = datetime.now()

            return RunResult(
                task_id=task.id,
                agent=self.agent,
                actual_diff=actual_diff,
                stdout=proc.stdout,
                stderr=proc.stderr,
                exit_code=proc.returncode,
                duration_seconds=elapsed,
                started_at=started_at,
                completed_at=completed_at,
                success=proc.returncode == 0,
            )

        except subprocess.TimeoutExpired:
            elapsed = time.monotonic() - start_time
            return RunResult(
                task_id=task.id,
                agent=self.agent,
                stderr="Agent timed out",
                exit_code=-1,
                duration_seconds=elapsed,
                started_at=started_at,
                completed_at=datetime.now(),
                success=False,
            )
        except Exception as e:
            elapsed = time.monotonic() - start_time
            return RunResult(
                task_id=task.id,
                agent=self.agent,
                stderr=str(e),
                exit_code=-1,
                duration_seconds=elapsed,
                started_at=started_at,
                completed_at=datetime.now(),
                success=False,
            )


def save_run_result(result: RunResult, output_dir: Path) -> Path:
    """Save a RunResult to JSON, return the file path."""
    output_dir.mkdir(parents=True, exist_ok=True)
    timestamp = result.started_at.strftime("%Y%m%d%H%M%S")
    filename = f"{result.task_id}-{result.agent}-{timestamp}.json"
    path = output_dir / filename
    path.write_text(result.model_dump_json(indent=2))
    return path

File: internal/runner/runner.go
// Package runner clones a repo at a specific commit, executes a coding agent,
// and captures the resulting git diff.
package runner

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"agenteval/internal/models"
)

// AgentConfigs maps well-known agent names to their command templates.
// {prompt} is replaced with the path to the prompt file.
// {workdir} is replaced with the repo working directory.
var AgentConfigs = map[string]string{
	"claude-code": "claude -p {prompt} --output-dir {workdir}",
	"aider":       "aider --message {prompt}",
}

// Runner executes a coding agent against a TaskDefinition.
type Runner struct {
	agent       string
	cmdTemplate string
	timeout     time.Duration
}

// New creates a Runner. If agentCmd is non-empty it overrides the built-in
// template for the named agent.
func New(agent, agentCmd string, timeout time.Duration) (*Runner, error) {
	var tmpl string
	if agentCmd != "" {
		tmpl = agentCmd
	} else if t, ok := AgentConfigs[agent]; ok {
		tmpl = t
	} else {
		known := make([]string, 0, len(AgentConfigs))
		for k := range AgentConfigs {
			known = append(known, k)
		}
		return nil, fmt.Errorf("unknown agent %q — known agents: %s — or pass --agent-cmd", agent, strings.Join(known, ", "))
	}
	return &Runner{agent: agent, cmdTemplate: tmpl, timeout: timeout}, nil
}

// RunTask clones the task's repo at base_commit into workDir, runs the agent,
// then returns the combined git diff.
func (r *Runner) RunTask(task models.TaskDefinition, workDir string) models.RunResult {
	startedAt := time.Now()

	result, err := r.runTask(task, workDir, startedAt)
	if err != nil {
		return models.RunResult{
			TaskID:          task.ID,
			Agent:           r.agent,
			Stderr:          err.Error(),
			ExitCode:        -1,
			DurationSeconds: time.Since(startedAt).Seconds(),
			StartedAt:       startedAt,
			CompletedAt:     time.Now(),
			Success:         false,
		}
	}
	return result
}

func (r *Runner) runTask(task models.TaskDefinition, workDir string, startedAt time.Time) (models.RunResult, error) {
	if err := os.MkdirAll(workDir, 0o755); err != nil {
		return models.RunResult{}, fmt.Errorf("creating workdir: %w", err)
	}

	repoDir := filepath.Join(workDir, "repo")

	// Clone
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
	defer cancel()
	if out, err := exec.CommandContext(ctx, "git", "clone", task.RepoURL, repoDir).CombinedOutput(); err != nil {
		return models.RunResult{}, fmt.Errorf("git clone: %w\n%s", err, out)
	}

	// Checkout base commit
	ctx2, cancel2 := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel2()
	if out, err := exec.CommandContext(ctx2, "git", "-C", repoDir, "checkout", task.BaseCommit).CombinedOutput(); err != nil {
		return models.RunResult{}, fmt.Errorf("git checkout %s: %w\n%s", task.BaseCommit, err, out)
	}

	// Write prompt file
	promptFile := filepath.Join(workDir, "prompt.txt")
	if err := os.WriteFile(promptFile, []byte(task.Prompt), 0o644); err != nil {
		return models.RunResult{}, fmt.Errorf("writing prompt file: %w", err)
	}

	// Build command
	cmd := strings.NewReplacer(
		"{prompt}", promptFile,
		"{workdir}", repoDir,
	).Replace(r.cmdTemplate)

	// Run agent
	agentCtx, agentCancel := context.WithTimeout(context.Background(), r.timeout)
	defer agentCancel()

	agentCmd := exec.CommandContext(agentCtx, "sh", "-c", cmd)
	agentCmd.Dir = repoDir
	agentOut, agentErr := agentCmd.CombinedOutput()

	stdout := string(agentOut)
	stderr := ""
	exitCode := 0
	if agentCmd.ProcessState != nil {
		exitCode = agentCmd.ProcessState.ExitCode()
	}
	if agentErr != nil {
		if agentCtx.Err() == context.DeadlineExceeded {
			stderr = "agent timed out"
			exitCode = -1
		} else {
			stderr = agentErr.Error()
		}
	}

	// Capture diff (staged + unstaged)
	diffCtx, diffCancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer diffCancel()
	diffOut, _ := exec.CommandContext(diffCtx, "git", "-C", repoDir, "diff").Output()
	cachedOut, _ := exec.CommandContext(diffCtx, "git", "-C", repoDir, "diff", "--cached").Output()
	actualDiff := string(diffOut) + string(cachedOut)

	return models.RunResult{
		TaskID:          task.ID,
		Agent:           r.agent,
		ActualDiff:      actualDiff,
		Stdout:          stdout,
		Stderr:          stderr,
		ExitCode:        exitCode,
		DurationSeconds: time.Since(startedAt).Seconds(),
		StartedAt:       startedAt,
		CompletedAt:     time.Now(),
		Success:         exitCode == 0,
	}, nil
}

// SaveRunResult writes result as JSON to outputDir and returns the file path.
func SaveRunResult(result models.RunResult, outputDir string) (string, error) {
	if err := os.MkdirAll(outputDir, 0o755); err != nil {
		return "", fmt.Errorf("creating output dir: %w", err)
	}
	timestamp := result.StartedAt.Format("20060102150405")
	filename := fmt.Sprintf("%s-%s-%s.json", result.TaskID, result.Agent, timestamp)
	path := filepath.Join(outputDir, filename)

	data, err := json.MarshalIndent(result, "", "  ")
	if err != nil {
		return "", err
	}
	return path, os.WriteFile(path, data, 0o644)
}
